api/src/main/java/org/pmiops/workbench/billing/BillingAlertsService.java

```diff
@@ -0,0 +1,80 @@
+package org.pmiops.workbench.billing;
+
+import com.google.cloud.bigquery.QueryJobConfiguration;
+import com.google.common.collect.Streams;
+import java.util.AbstractMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.stream.Collectors;
+import javax.inject.Provider;
+import org.pmiops.workbench.api.BigQueryService;
+import org.pmiops.workbench.config.WorkbenchConfig;
+import org.pmiops.workbench.db.dao.UserDao;
+import org.pmiops.workbench.db.model.User;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class BillingAlertsService {
+
+  private final BigQueryService bigQueryService;
+  private final NotificationService notificationService;
+  private final UserDao userDao;
+  private final Provider<WorkbenchConfig> workbenchConfigProvider;
+
+  @Autowired
+  public BillingAlertsService(
+      BigQueryService bigQueryService,
+      NotificationService notificationService,
+      UserDao userDao,
+      Provider<WorkbenchConfig> workbenchConfigProvider) {
+    this.bigQueryService = bigQueryService;
+    this.notificationService = notificationService;
+    this.userDao = userDao;
+    this.workbenchConfigProvider = workbenchConfigProvider;
+  }
+
+  public void alertUsersExceedingFreeTierBilling() {
+    QueryJobConfiguration queryConfig =
+        QueryJobConfiguration.newBuilder(
+                "SELECT project.id, SUM(cost) cost FROM `"
+                    + workbenchConfigProvider.get().billing.billingExportBigQueryTable
+                    + "` GROUP BY project.id ORDER BY cost desc;")
+            .build();
+
+    Map<User, Double> perUserSpend =
+        Streams.stream(bigQueryService.executeQuery(queryConfig).getValues())
+            .filter(fv -> !fv.get("id").isNull() && !fv.get("id").isNull())
+            .map(
+                fv ->
+                    new AbstractMap.SimpleImmutableEntry<>(
+                        userDao.findCreatorByWorkspaceNamespace(fv.get("id").getStringValue()),
```

Comment:
I think it will probably be easier to code and will scale better to instead start before this stream with a single map of  BP -> user. Ultimately, you'll likely need to fetch every user.